interface/src/app/analyze/analysis/sampleBin.js

```diff
@@ -277,15 +280,65 @@ function($log, $cacheFactory, $q, Sample, Activity) {
       //  note: progress can be reported by returning a $promise to the caller
       // respObj.queryStatus = 'Retrieving sample activity...';
       this.rebuildHeatmapActivity(this.heatmapData.samples);
+    },
+
+    // volcano plot methods
+    getVolcanoPlotData: function() {
+      // use sample lists for group-a and group-b to produce output for
+      // the volcano plot of the form:
+      //   node - diff - logsig,
+      // where:
+      //   node = the node name as supplied by NodeInfo
+      //   diff = mean(group-a activity values) - mean(group-b activity values)
+      //   logsig = -log10(p-value from 2-sample t-test on group-a vs. group-b)
+      var sg = this.getSamplesByGroup();
+
+      // verify that we have at least one sample each in group-a and group-b
+      if (!sg['group-a'] || sg['group-a'].length === 0) {
+        return null;
+      }
+      if (!sg['group-b'] || sg['group-b'].length === 0) {
+        return null;
+      }
+
+      // (1) we obtain a list of nodes by retrieving node activity
+      //     for the first sample in our volcano plot
+      var firstSampleNodes = this.activityCache.get(sg['group-a'][0]);
+      // (2a) next, we build a new array (`retval`) comprised of `nodeObject`s
+      //      by walking through the `firstSampleNodes` and constructing a
+      //      `nodeObject` for each. [outer .map()]
+      var retval = firstSampleNodes.map(function(val, i, arr) {
+        var mapSampleIdsToActivity = function(sampId, i, arr) {
+          // (2b) the array of activity for each node is built by plucking the
+          //      activity `.value` for each sample within this node from the
+          //      `activityCache` [inner .map()]
+          // FIXME: counting on array order to match node order here
+          return this.activityCache.get(sampId)[val.node - 1].value;
+        };
+        var nodeObject = {
+          'id': val.node, // TODO: map this id to node name (via vega spec?)
+          'activityA': sg['group-a'].map(mapSampleIdsToActivity, this),
+          'activityB': sg['group-b'].map(mapSampleIdsToActivity, this)
+        };
+        nodeObject.diff = (
+          MathFuncts.mean(nodeObject.activityA) -
+          MathFuncts.mean(nodeObject.activityB)
+        );
+        nodeObject.logsig = -Math.log10(MathFuncts.tTest(
+          nodeObject.activityA, nodeObject.activityB
+        ).pValue());
+
+        return nodeObject;
+      }, this);
+      return retval;
     }
   };
 
   return SampleBin;
```

Comment:
Never mind, using `service` will save a few lines, but if you are not familiar with it, forget it.